added server.FolderWatchList to cli + LocalSourcesWatcher

akramsystems opened the pull request:
## Describe your changes

This PR adds a new configuration option `server.folderWatchList` to allow users to specify a custom path to watch for file changes. Currently, Streamlit only supports watching the directory of the main script or excluding files via a blacklist, but this change provides the flexibility to monitor external directories or specific folders. Additionally, logic has been added to ensure the new path is recursively watched for file changes, and tests have been updated to cover this feature.

Note: This is my first time doing a PR to open source, so any feedback would be appreciated.

## GitHub Issue Link (if applicable)

Issue Number: #9655 

## Testing Plan

- Unit tests were added to `local_sources_watcher_test.py` to ensure the custom watch path feature behaves as expected.

### MANUAL TEST
`streamlit run app/hello.py --server.runOnSave=true --server.folderWatchList=<BASE_PATH>/test_folder --logger.level=debug`

```
...
...
├── app/
│   ├── hello.py
└── test_folder
    └── test_file.txt
```

**app/hello.py**
```app/hello.py
import streamlit as st

st.write("Hello, World!")

with open("test_folder_1/test_file.txt", "r") as f:
    st.write("test_folder_1/test_file.txt")
    st.write(f.read())

with open("test_folder_2/test_file.txt", "r") as f:
    st.write("test_folder_2/test_file.txt")
    st.write(f.read())
```

**.streamlit/config.toml**
```.streamlit/config.toml
[server]
folderWatchList = [
    "test_folder_1",
    "test_folder_2"
]
runOnSave = true
```

**Note**: the list of folders should be _absolute paths_ currently, thought this is a good first step and then we can add regex search or relative paths in another PR if possible, since smaller PR's are more manageable/




**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

akramsystems committed:
added server.customWatchPath as cli param to click and added all files under the folder to PathWatcher for the LocalSourcesWatcher

jrieke commented:
Hey, this is awesome, thanks for the PR! We wanted something like this for a long time. I think it would be great to align this a bit more to the existing `folderWatchBlacklist` by a) accepting a list of folders instead of just a single path and b) renaming the parameter to `folderWatchList` (we could also do `folderWatchWhitelist` but I think that's unnecessarily complicated). What do you think?

akramsystems commented:
@jrieke for sure I'll implement that today, sorry I was busy during the week!  I can definitely do that, and yes i agree using the name _folderWatchWhitelist_ is too long and unnecessary.

akramsystems committed:
gh-9655 set config option name to folderWatchList and now accepting a list of folder paths instead of a single path to watch

akramsystems commented:
@jrieke updated the PR take a look note i removed the need to add the custom_watch_path to the _EventsBasedWatcher_ since that code wasn't needed or used .

akramsystems committed:
gh-9965 rm changes to event_based_path_watcher

jrieke commented:
Sweet! I just triggered the test runs, let's see if there are any issues. And I also pinged our engineers to give this a proper review.

akramsystems commented:
> And I also pinged our engineers to give this a proper review.

Awesome I'll await their feedback, also I update the branch when ever it is behind which requires the test to be run again but i seen that 22 passed and there were no errors.  Feel free to trigger it again if you like, and thanks again for being proactive with this PR

kmcgrady commented on lib/streamlit/watcher/local_sources_watcher.py:
```diff
@@ -79,6 +80,21 @@ def update_watched_pages(self) -> None:
+
+        for watch_folder in self._watch_folders:
+            _LOGGER.debug("Registering watch folder: %s", watch_folder)
+            if os.path.exists(watch_folder):
```
We probably should be validating that the folder is a directory otherwise the `is_directory` will be not valid.

akramsystems replied:
for sure will add this check

kmcgrady commented on lib/streamlit/watcher/local_sources_watcher.py:
```diff
@@ -132,10 +149,20 @@ def _register_watcher(self, filepath, module_name):
-                module_name=module_name,
-            )
+            if is_directory:
+                for root, _, files in os.walk(filepath):
```
We do have logic that walks a full directory. https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/watcher/path_watcher.py#L143

It seems to just create a watch on the path, and that can allow for . https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/watcher/path_watcher.py#L119

Perhaps we can just do a watch on the path itself. Perhaps we can modify `_register_watcher`. We can test if the glob pattern needs to be set. This would reduce watchers generated and perhaps simplify the logic.

akramsystems replied:
for sure i'll look at incorporating the `watch_dir` function and try to use the glob pattern to make watching paths easier

kmcgrady reviewed (commented):
Hey @akramsystems! Thank you for creating a PR. I realized my review didn't send out, so apologies for the delay. Overall things are looking good, I provided a couple comments on where we can tighten things up and potentially reduce memory/process overhead. Would you mind taking a look and let me know if it helps us?

akramsystems committed:
gh-9655 checking if watch_folders are indeed directories + switched to using watch_dir with glob_pattern to easily walk directory + using a new callback to trigger changes within directories specifically + updated events based path watcher to take account for triggering and event to fire the callback function when a directory is changed + updated test to reflect new logic

akramsystems commented on lib/streamlit/watcher/local_sources_watcher.py:
```diff
@@ -79,6 +81,21 @@ def update_watched_pages(self) -> None:
+
+        for watch_folder in self._watch_folders:
+            # check if it is folder
+            if not os.path.isdir(watch_folder):
```
as requested i'm checking if the the paths are directories

akramsystems commented on lib/streamlit/watcher/local_sources_watcher.py:
```diff
@@ -92,7 +109,23 @@ def update_watched_pages(self) -> None:
     def register_file_change_callback(self, cb: Callable[[str], None]) -> None:
         self._on_file_changed.append(cb)
 
+    def on_dir_changed(self, filepath: str):
```
Created a seperate function for this as I didn't want to interfere with existing logic which uses the on_file_changed callback, and make it more explicit that this callback can be called for changes to directories

kmcgrady replied:
I didn't need to create this function.

akramsystems replied:
removed this function

akramsystems commented on lib/streamlit/watcher/local_sources_watcher.py:
```diff
@@ -132,10 +165,20 @@ def _register_watcher(self, filepath, module_name):
-            )
+            if is_directory:
+                # Watch the directory instead of individual files
+                watch_dir(filepath, self.on_dir_changed, glob_pattern="**/*")
```
as requested using the watch_dir with the usage of `glob_pattern` to search all subdirectories and types of files to walk the directory files

kmcgrady replied:
I think this will simplify if you update the entire try block to just be this.

```python
kwargs = {}
if is_directory:
    kwargs["glob_pattern"] = "**/*"

wm = WatchedModule(
    watcher=PathWatcher(filepath, self.on_file_changed, **kwargs),
    module_name=module_name,
)
self._watched_modules[filepath] = wm
```

Apologies if I was confusing, but my comment was on `watch_dir` was to borrow the logic from it. Looking at the function, it's simply creating a watcher with the glob pattern, and we already have the watcher, so we just need to specify the glob pattern if it's a directory. `self.on_file_changed` will have the correct file path because the watcher will supply the directory path, and not the actual file that changed. That's fine with us because we just want to rerun the page if we detect _any_ change.

As an additional cleanup, I might consider modifying the name of `on_file_changed` to `on_path_changed`.

akramsystems replied:
applied this

akramsystems commented on lib/streamlit/watcher/event_based_path_watcher.py:
```diff
@@ -360,6 +360,14 @@ def handle_path_change_event(self, event: events.FileSystemEvent) -> None:
         abs_changed_path = os.path.abspath(changed_path)
 
         changed_path_info = self._watched_paths.get(abs_changed_path, None)
+        if changed_path_info is None:
```
This needed to be done since the event_based_path_watcher only checks for abs_changed_paths, this approach checks if the path that has changed is a common ancestor of the parent path. making it simple to check for for changes to a directory no matter the file under the directory

kmcgrady replied:
So I was able to get everything working with removing this code. Seems to work fine in my tests.

akramsystems replied:
@kmcgrady I can remove the check for `if changed_path_info is None`, but if I remove the for loop going over the `watched_paths.items()`and checking if they share common path then it fails to update the app there is a change in the directories being watched.

akramsystems reviewed (commented):
@kmcgrady I made some small changes (and left some comments) let me know your thoughts, or if this was what you had in mind.  Your feedback is greatly appreciated! 👍🏽

kmcgrady commented on lib/streamlit/config.py:
```diff
@@ -580,6 +580,21 @@ def _logger_message_format() -> str:
+        Note: This is a list of absolute paths.
+    """,
+    default_val=[],
+    type_=list,
```
You shouldn't need to specify the `type_`. I think it's a little weird, but I think it should be str because the values are strings, but we will accept multiple strings for the paths. The fact that the default val is a list hints at taking multiple values, and str is inferred by default.

kmcgrady reviewed (commented):
Apologies for the delay. I'm working hard to wrap up some stuff before the holidays occur.

I provided some comments with suggested changes. I think we can simplify this a lot more. Can you check to see if the changes work for your use cases. I did some testing of editing files in a directory, and having the script rerun successfully for both Event and Polling Strategy.

I am around this week if you want to talk more. After that, I go on vacation, and will resume reviewing in the new year.

syasini committed:
Fix Typo in README.md (#10056)

Trying to set a record for the smallest code contribution to Streamlit. 
In an ironic twist, I completely ignored the very instructions I'm
updating and skipped creating an issue first.

## Describe your changes
Fixing a typo in the README.md file: 
Ff -> If.
Tiny change. Big impact. 

## GitHub Issue Link (if applicable)
Taking the scenic route and skipping the issue for this monumental PR.

## Testing Plan

- Visual inspection by reviewer.
- Reflecting on the irony of fixing instructions I did not follow...

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

kmcgrady commented:
Hey @akramsystems Just following up to ask if you plan to continue working on this PR and if the feedback given makes sense. Let me know! :-)

akramsystems commented:
@kmcgrady yes I plan on continuing  to work on this PR, sorry i had a surgery in the end of the year so that caused me to delay some of my work.  Will continue the grind! Yes your comments make sense i will update the PR

kmcgrady commented:
I hope your surgery went well and a speedy recovery. Let me know how I can help!

kmcgrady commented:
Hey @akramsystems I updated the title with `[WIP]` for Work in Progress. It provides a way for us to filter PRs awaiting reviews. Once you have made changes and feel ready for review, feel free to just remove the [WIP] and it will pop back on my list to review.

N-Demir commented:
+1 this is an awesome addition!

eliasecchig commented:
+1 on this!

akramsystems committed:
gh-9655 rm type list from config (address pr comment)

akramsystems committed:
gh-9655 Refactor file watcher to simplify directory watching implementation

akramsystems commented:
@kmcgrady  ready for re-review I applied all but one of your recommendations

akramsystems committed:
gh-9655 rm check for changed_info is None

kmcgrady commented on lib/streamlit/watcher/event_based_path_watcher.py:
```diff
@@ -360,6 +360,13 @@ def handle_path_change_event(self, event: events.FileSystemEvent) -> None:
         abs_changed_path = os.path.abspath(changed_path)
 
         changed_path_info = self._watched_paths.get(abs_changed_path, None)
+        for path, info in self._watched_paths.items():
```
Might benefit from a comment that explains what this is doing. Something like:

```python
# Check if any of our watched paths includes the changed path and is actually being watched
```

akramsystems replied:
added a comment

kmcgrady reviewed (commented):
Things are looking food. I think you just need to satisfy the type checkers (and I just requested a comment).

You can run the typechecker with `make mypy`

akramsystems committed:
gh-9655 added comment explaning how files in folders are being watched + fixed mypy issues

akramsystems commented:
@kmcgrady I fixed the mypy errors on my code changes but there were some mypy errors unrelated to the changes I made, I'll leave them as is?

kmcgrady commented:
@akramsystems Seems like mypy passed just fine, but there seems to be a python test failure.

akramsystems committed:
gh-9655 fix tests (ran make pytests to verify)

akramsystems commented:
@kmcgrady seems like the tests are passing ?

kmcgrady commented:
Great work @akramsystems Approved and merged

akramsystems commented:
HALLALUYAH THANK YOU SO MUCH @kmcgrady

jrieke commented:
Wohoo, congrats for getting this merged @akramsystems!

sfc-gh-jesmith commented:
Awesome @akramsystems! 
Would love to send ya a lil Streamlit swag to say thanks..feel free to [fill out this form](https://forms.gle/RZCUTKCrg7aHyqH78) if interested. 😊

adamalton commented:
@sfc-gh-jesmith it looks like this was merged to `develop` on 2025/04/28 ([merge commit](https://github.com/streamlit/streamlit/commit/d9c9c998d1bc14f4c81991ed0c0b7fc363f4e8a4)), and then the next day the [1.45.0 release](https://github.com/streamlit/streamlit/releases/tag/1.45.0) happened, but didn't contain the change, and then 2 weeks later on 2025/05/12 [release 1.45.1](https://github.com/streamlit/streamlit/releases/1.45.1) happened, but that also didn't contain this change.

I don't think it's just an issue with the release notes, as doing `git tag --contains d9c9c998d1bc14f4c81991ed0c0b7fc363f4e8a4` (the merge commit) gives:
```     
1.44.2.dev20250428
1.45.1.dev20250429
1.45.1.dev20250430
1.45.1.dev20250501
1.45.1.dev20250502
1.45.1.dev20250503
1.45.1.dev20250504
1.45.1.dev20250505
1.45.1.dev20250506
1.45.1.dev20250507
1.45.1.dev20250508
1.45.1.dev20250509
1.45.1.dev20250510
1.45.1.dev20250511
1.45.2.dev20250512
1.45.2.dev20250513
1.45.2.dev20250514
1.45.2.dev20250515
1.45.2.dev20250516
1.45.2.dev20250517
1.45.2.dev20250518
1.45.2.dev20250519
1.45.2.dev20250520
1.45.2.dev20250521
1.45.2.dev20250522
1.45.2.dev20250523
1.45.2.dev20250525
1.45.2.dev20250526
```

Only dev tags 🤷 .

The change definitely exists in the `develop` branch, but if I download the source code for the 1.45.1 release, the change is definitely not in there.

Do you know what's happened to this change, and/or when it's going to make it into a release? Thanks! 🙏

lukasmasuch commented:
@adamalton the change will be released in 1.46

adamalton commented:
@lukasmasuch Thank you 🙏

akramsystems commented:
> Awesome @akramsystems! Would love to send ya a lil Streamlit swag to say thanks..feel free to [fill out this form](https://forms.gle/RZCUTKCrg7aHyqH78) if interested. 😊

Hey sorry for taking a while getting back to you, but yes I just replied to the form :D !
